src/rgw/rgw_d3n_datacache.cc

```diff
@@ -0,0 +1,328 @@
+  if (it != outstanding_write_list.end()) {
+    cache_lock.unlock();
+    ldout(cct, 10) << "D3nDataCache: Warning: data put in cache already issued, no rewrite" << dendl;
+    return;
+  }
+  outstanding_write_list.push_back(oid);
+  cache_lock.unlock();
+
+  eviction_lock.lock();
+  _free_data_cache_size = free_data_cache_size;
+  _outstanding_write_size = outstanding_write_size;
+  eviction_lock.unlock();
+
+  ldout(cct, 20) << "D3nDataCache: Before eviction _free_data_cache_size:" << _free_data_cache_size << ", _outstanding_write_size:" << _outstanding_write_size << ", freed_size:" << freed_size << dendl;
+  while (len >= (_free_data_cache_size - _outstanding_write_size + freed_size)) {
+    ldout(cct, 20) << "D3nDataCache: enter eviction, r=" << r << dendl;
+    if (eviction_policy == _eviction_policy::LRU) {
+      r = lru_eviction();
+    } else if (eviction_policy == _eviction_policy::RANDOM) {
+      r = random_eviction();
+    } else {
+      ldout(cct, 0) << "D3nDataCache: Warning: unknown cache eviction policy, defaulting to lru eviction" << dendl;
+      r = lru_eviction();
+    }
+    if (r < 0)
+      return;
+    freed_size += r;
+  }
+  r = create_aio_write_request(bl, len, oid);
+  if (r < 0) {
+    cache_lock.lock();
+    outstanding_write_list.remove(oid);
+    cache_lock.unlock();
+    ldout(cct, 1) << "D3nDataCache: create_aio_write_request fail, r=" << r << dendl;
+    return;
+  }
+
+  eviction_lock.lock();
+  free_data_cache_size += freed_size;
+  outstanding_write_size += len;
+  eviction_lock.unlock();
+}
+
+bool D3nDataCache::get(const string& oid)
+{
+  bool exist = false;
+  string location = cache_location + oid;
+
+  lsubdout(g_ceph_context, rgw_datacache, 20) << "D3nDataCache: " << __func__ << "(): location=" << location << dendl;
+  cache_lock.lock();
+  map<string, D3nChunkDataInfo*>::iterator iter = cache_map.find(oid);
+  if (!(iter == cache_map.end())) {
+    // check inside cache whether file exists or not!!!! then make exist true;
+    struct D3nChunkDataInfo* chdo = iter->second;
+    if (access(location.c_str(), F_OK ) != -1 ) { // file exists
+      exist = true;
+      {
+        /*LRU*/
+        /*get D3nChunkDataInfo*/
+        eviction_lock.lock();
+        lru_remove(chdo);
+        lru_insert_head(chdo);
+        eviction_lock.unlock();
+      }
+    } else {
+      cache_map.erase(oid);
+      lru_remove(chdo);
+      exist = false;
+    }
+  }
+  cache_lock.unlock();
+  return exist;
+}
+
+size_t D3nDataCache::random_eviction()
+{
+  lsubdout(g_ceph_context, rgw_datacache, 20) << "D3nDataCache: " << __func__ << "()" << dendl;
+  int n_entries = 0;
+  int random_index = 0;
+  size_t freed_size = 0;
+  D3nChunkDataInfo* del_entry;
+  string del_oid, location;
+
+  cache_lock.lock();
+  n_entries = cache_map.size();
+  if (n_entries <= 0) {
+    cache_lock.unlock();
+    return -1;
+  }
+  srand (time(NULL));
+  random_index = rand()%n_entries;
```

Comment:
don't change the seed at runtime

ceph has utilities based on the c++ `<random>` library. you can just use:
```c++
ceph::util::generate_random_number<int>(0, n_entries-1);
```

Reply:
> other code in rgw sends http requests with RGWHTTPClient and RGWHTTPManager, which uses the curl_multi interface to do this with a single thread. we'll want to use that instead

yes, copied the comment to the l2 pr (#38748)